### PR TITLE
a more elegant way about HashedWheelTimer

### PR DESCRIPTION
**Pull request by CalvinKirs**
Motivation:

Explain here the context, and why you're making that change.
What is the problem you're trying to solve.

I think it might be better to do this. When ticksPerWheel  its big , the efficiency will be obvious

Modification:

Describe the modifications you've done.

Result:
Using binary bit operations

Fixes #<GitHub issue number>. 

If there is no issue then describe the changes introduced by this PR.

no

### CODE REVIEW

**Comment by CalvinKirs** [comment]
> Can one of the admins verify this patch?

of course

**Comment by carl-mastrangelo** [comment]
hi @CalvinKirs 

Would you mind running the benchmarks to show if there is a performance difference?

**Comment by CalvinKirs** [comment]
> hi @CalvinKirs
> 
> Would you mind running the benchmarks to show if there is a performance difference?

sure,I have added a simple test case to prove that my approach does improve efficiency, you can look

**Comment by CalvinKirs** [comment]
> Please use a proper benchmark eg JMH to perform test to validate improvementes

I have finished, please check

**Comment by CalvinKirs** [comment]
> Please use a proper benchmark eg JMH to perform test to validate improvementes

I have finished, please check

**Comment by slandelle** [comment]
@CalvinKirs As @johnou asked, could you please revert all your formatting changes (check https://github.com/netty/netty/pull/9188/files) as those are related to your actual PR.

**Comment by CalvinKirs** [comment]
> @CalvinKirs As @johnou asked, could you please revert all your formatting changes (check https://github.com/netty/netty/pull/9188/files) as those are related to your actual PR.



> @CalvinKirs正如@johnou问的那样，您能否请将所有格式更改还原（请查看https://github.com/netty/netty/pull/9188/files），因为这些更改与您的实际PR相关。

so sory，My ide code formatting is different, I have completed the change

**Comment by CalvinKirs** [comment]
> It would be better to extend some of the other existing JMH benchmarks in Netty.
> Anyway, which results this one is having for you?
> It would be good to benchmark the effects of this change into HashedWheelTimer too.



> It would be better to extend some of the other existing JMH benchmarks in Netty.
> Anyway, which results this one is having for you?
> It would be good to benchmark the effects of this change into HashedWheelTimer too.

Ok, wait a moment.thank you

**Comment by CalvinKirs** [comment]
> It would be better to extend some of the other existing JMH benchmarks in Netty.
> Anyway, which results this one is having for you?
> It would be good to benchmark the effects of this change into HashedWheelTimer too.

I've completed

**Comment by slandelle** [comment]
Your JMH benchmarks are most likely broken because your computation isn't used and might get JIT'ed away. Your computation result must be either returned, or fed into an explicit BlackHole. Please read https://www.oracle.com/technetwork/articles/java/architect-benchmarking-2266277.html.

File: common/src/main/java/io/netty/util/HashedWheelTimer.java [hunk 1]
```diff
@@ -323,11 +323,15 @@ protected void finalize() throws Throwable {
     }
 
     private static int normalizeTicksPerWheel(int ticksPerWheel) {
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: No, no, when the amount of HashedTimerWheel is large enough, the number of calls to the method will also increase, and in netty, the amount of a single HashedTimerWheel will be large. Another point, we always want our code to be the simplest and the best, right?

**Comment by CalvinKirs** [comment]
> Your JMH benchmarks are most likely broken because your computation isn't used and might get JIT'ed away. Your computation result must be either returned, or fed into an explicit BlackHole. Please read https://www.oracle.com/technetwork/articles/java/architect-benchmarking-2266277.html.

I have finished, please check

File: common/src/main/java/io/netty/util/HashedWheelTimer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnou) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: That's my point, you should share HashedTimerWheel, not instantiate it many times.

[the remark this answers is not in the document]
Comment (CalvinKirs) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: Of course, but this method is triggered every time a task is added.

[the remark this answers is not in the document]
Comment (CalvinKirs) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: > That's my point, you should share HashedTimerWheel, not instantiate it many times.

Of course, but this method is triggered every time a task is added.

[the remark this answers is not in the document]
Comment (CalvinKirs) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: > That's my point, you should share HashedTimerWheel, not instantiate it many times.

Yes, Really bring about a slight increase in performance，then, can the PR be closed?

[the remark this answers is not in the document]
Comment (CalvinKirs) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: > > That's my point, you should share HashedTimerWheel, not instantiate it many times.
> 
> Yes, Really bring about a slight increase in performance，then, can the PR be closed?

But this way is more elegant, what do you think?

[the remark this answers is not in the document]
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: Honestly I think the code is fine as it is...

[the remark this answers is not in the document]
Comment (CalvinKirs) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: > Honestly I think the code is fine as it is...

Do you mean old or new?I sincerely hope to get your advice.

[the remark this answers is not in the document]
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/HashedWheelTimer.java]: The old implementation...

**Pull request closed without merging**